qp/utils.py

```diff
@@ -0,0 +1,46 @@
+import numpy as np
+import sys
+
+def safelog(arr, threshold=sys.float_info.epsilon):
+    """
+    Takes log of array with zeroes.
+
+    Parameters
+    ----------
+    arr: ndarray
+        Values to be logged
+    threshold: float
+        Small, positive value to replace zeros and negative numbers
+
+    Returns
+    -------
+    logged: ndarray
+        Logged values, with approximation in place of zeros and negative numbers
+    """
+    shape = np.shape(arr)
+    flat = arr.flatten()
+    logged = np.log(np.array([max(a,threshold) for a in flat])).reshape(shape)
+    return logged
+
+def calckl(qn, pn):
+    """
+    Calculates Kullback-Leibler Divergence
+
+    Parameters
+    ----------
+    qn: ndarray
```

Comment:
Good. This brings up a more general point. Let's imagine a bright future where some enlightened photo-z package produces catalogs of galaxy redshifts where the posterior PDF for each one is represented  with either 100 evenly spaced quantiles (in catalog Q), or 100 evenly spaced histogram bins (in catalog P). How would we use `qp` to compare the two options? For the kth galaxy we could read its Q entry into a `qp.PDF` object `Q` and its P entry into a `qp.PDF` object `P`, and then ask for `KL(P,Q)`. For this to work, we need the PDF objects to use their `approximate` method if their `truth` is `None` - and we need the `approximate` method to use the `quantiles` if it has them and the histogram bins otherwise (or throw an error if it has none of `truth`, `quantiles` or `bins`.  One thing I like about this way of thinking is that it forces us to admit that oftentimes we _say_ we are using the "full PDF" when actually we are using an approximate characterization of it. What do you think?